In [1]:
import numpy as np
import pandas as pd

df=pd.read_csv("/home/sajid/house-price-prediction/data/train.csv")

In [2]:
#handle missing values

In [3]:
#if numerical column,fill the nan values with the median of that column .
# the reason of using median over mean is that in the EDA,we came to know that the saleprice is right skewed so the outliers it pulls the value of mean upward
# so if data is skewed mean gets distorted but the median stills remain near the central tendency

In [4]:
num_cols = df.select_dtypes(include=np.number).columns

for col in num_cols:
    df[col] = df[col].fillna(df[col].median())

In [5]:
#if categorical column,fill 

In [6]:
# fill the categorical missing values with the mode ie the most repeating category

In [7]:
cat_cols = df.select_dtypes(exclude=np.number).columns

for col in cat_cols:
    df[col] = df[col].fillna(df[col].mode()[0])

In [8]:
#one hot encoding : the linear regression model only understands the numerical values. so converting the categorial coumns into binary 0/1 columns

In [9]:
df = pd.get_dummies(df, drop_first=True)

In [10]:
# if somehow the creation of dummy columns lead to multicollinearity,that can cause inverse(Xtranspose.X) to be singular
#so drop first drops the first column and avoids that 

In [11]:
df=df.astype(float)

In [12]:
X = df.drop("SalePrice", axis=1).values

y = df["SalePrice"].values

In [13]:
from sklearn.model_selection import train_test_split


In [14]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=42)

In [15]:
from my_library.preprocessing import StandardScaler

In [16]:
scaler=StandardScaler()

In [17]:
X_train=scaler.fit_transform(X_train)
X_test=scaler.transform(X_test)

In [18]:
from my_library.multiple_linear_regression_closed_form import multiple_linear_regression_closed_form

In [19]:
model=multiple_linear_regression_closed_form()

In [20]:
model.fit(X_train,y_train)

In [21]:
y_pred=model.predict(X_test)

In [22]:
from my_library.metrics import r2_score

In [23]:
print(r2_score(y_test,y_pred))

0.6554927583866419


# Model Evaluation

- Multiple Linear Regression was implemented using the closed form normal equation.
- Features were standardized before training.
- The model achieved an R² score of approximately 0.655.
- The model explains around 65% of the variance in house prices.